### PR TITLE
Fix windows debug build

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
Fix a small number of issues which were preventing the unit tests from passing on windows-debug builds.

Closes https://github.com/scipp/scipp/issues/2055

### CODE REVIEW

File: variable/test/transform_test.cpp [hunk 1]
```diff
@@ -772,9 +772,10 @@ TEST_F(TransformInPlaceBucketsDryRunTest, unchanged_if_success) {
 TEST(TransformFlagsTest, no_variance_on_arg) {
   auto var_with_variance = makeVariable<double>(Values{1}, Variances{1});
   auto var_no_variance = makeVariable<double>(Values{1});
-  auto binary_op = [](auto x, auto y) { return x + y; };
-  auto op_arg_0_has_flags =
+  constexpr auto binary_op = [](auto x, auto y) { return x + y; };
```
Comment (jl-wynen) [inline variable/test/transform_test.cpp:775]: Looks like this relates to the error that those variables were used uninitialised. Do you know why `constexpr` fixes it?

Comment (Tom-Willemsen) [reply variable/test/transform_test.cpp:775]: In all honesty, no, I'm not sure why adding `constexpr` fixes it. From https://docs.microsoft.com/en-us/cpp/cpp/lambda-expressions-constexpr?view=msvc-160 I believe this lambda would have been "implicitly constexpr" anyway. The line below with the `scipp::overloaded` call (also changed to `constexpr`) is where it was actually triggering the breakpoint (not sure how accurate this location really is). Everything looks like it was initialized before being used to me... and if it wasn't, it shouldn't compile with constexpr... unless there is some subtelty I've missed with `scipp::overloaded` or one of the `transform_flags`.

The other uses of `scipp::overloaded` in scipp to all use either `constexpr` or use/return the result immediately.

Comment (jl-wynen) [reply variable/test/transform_test.cpp:775]: `overloaded` only cares about the type, it does not use the instance of the lambda you pass in at all. Maybe it has to do with how the `overloaded` objects' parents are constructed.

But as you say, if it were truly uninitialised, it shouldn't compile.

I will just keep in mind to make all transform kernel `constexpr`

**Comment by Tom-Willemsen** [comment]
Note: currently expecting the new debug pipeline to fail, until the `simpler-index` branch is merged, which fixes an array overrun in `MultiIndexTest.6d`.

**Comment by SimonHeybrock** [comment]
@Tom-Willemsen Can fix this PR (or close if not necessary any more)? Cheers!

**Comment by Tom-Willemsen** [comment]
As far as I can tell the `simpler-index` stuff [here](https://github.com/scipp/scipp/pull/1900) which this depends on hasn't been merged yet. Once that PR is merged I can re-check whether it fixes the issue seen here.

File: CMakeLists.txt [hunk 2]
```diff
@@ -190,6 +190,10 @@ if(NOT WIN32)
     $<$<COMPILE_LANGUAGE:CXX>:-Woverloaded-virtual>
     $<$<COMPILE_LANGUAGE:CXX>:-fno-operator-names>
   )
+else()
+  add_compile_options(
+    $<$<COMPILE_LANGUAGE:CXX>:/W4> $<$<COMPILE_LANGUAGE:CXX>:/bigobj>
```
Comment (jl-wynen) [inline CMakeLists.txt:195]: What does `/bigobj` do? And why did we not have it before?

File: tools/build_cpp.py [hunk 3]
```diff
@@ -62,8 +73,12 @@ def main(*,
         '-DWITH_CTEST': 'OFF',
         '-DCMAKE_INTERPROCEDURAL_OPTIMIZATION': ipo,
         '-DFULL_BUILD': 1,
+        '-DCMAKE_BUILD_TYPE': 'Debug' if debug_build else "Release",
     }
 
+    if debug_build:
+        cmake_flags["-DDYNAMIC_LIB"] = "ON"
```
Comment (jl-wynen) [inline tools/build_cpp.py:80]: Do we have to enforce this or can MSVC also make a debug build with static libraries?

File: .azure-pipelines/debug_builds.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
+parameters:
```
Comment (SimonHeybrock) [inline .azure-pipelines/debug_builds.yml:1]: Is this file still needed?

File: CMakeLists.txt (hunk 2 above)
Comment (Tom-Willemsen) [reply CMakeLists.txt:195]: Bigobj increases the number of objects you can store in a `.obj` file (https://docs.microsoft.com/en-us/cpp/build/reference/bigobj-increase-number-of-sections-in-dot-obj-file?view=msvc-160). The build was warning about this previously.

File: .azure-pipelines/debug_builds.yml (hunk 4 above)
Comment (Tom-Willemsen) [reply .azure-pipelines/debug_builds.yml:1]: If we still want to set up a CI pipeline for windows-debug builds, yes. If we've decided that we don't want a pipeline for this, then we can remove this.

Comment (SimonHeybrock) [reply .azure-pipelines/debug_builds.yml:1]: What would be the use of the pipeline?

Comment (Tom-Willemsen) [reply .azure-pipelines/debug_builds.yml:1]: To spot the kinds of issues fixed in this PR soon after they get merged, rather than waiting for a windows developer to hit them when trying to build in debug mode.

Comment (SimonHeybrock) [reply .azure-pipelines/debug_builds.yml:1]: If we wanted to do that, couldn't we just us a parameter on one of the existing pipelines, rather than maintaining a separate yml that will eventually break?

Comment (Tom-Willemsen) [reply .azure-pipelines/debug_builds.yml:1]: I think we decided we didn't want to do that as it would then build on every commit, using up more azure resources than we wanted. This is set to build nightly only. If we've changed our mind on this I'm happy to adjust it to use a parameter on existing pipelines.

Comment (SimonHeybrock) [reply .azure-pipelines/debug_builds.yml:1]: I am not suggesting to adjust any parameters, rather, if we want to make a debug build, use a parameter of the existing pipelines.

Comment (Tom-Willemsen) [reply .azure-pipelines/debug_builds.yml]: As far as I can tell from the azure docs it's not possible to use two different scheduling settings (i.e. cron and pr triggers) to schedule builds with different parameters in a single `.yml` file. My understanding from the docs is that they want you to use 2 separate `.yml` files for this (which might share a template). Since you were concerned this morning about using two `.yml` files I have now changed it to use parameters in the existing `.yml`, but this does mean our windows-debug builds will be triggered at the same time as all our other builds (i.e. for commits in every PR)

Also depends on the pipelines changes at https://github.com/scipp/pipelines/pull/22

Comment (SimonHeybrock) [reply .azure-pipelines/debug_builds.yml]: > but this does mean our windows-debug builds will be triggered at the same time as all our other builds (i.e. for commits in every PR)

As discussed previously, we cannot do that, since we are running out of builders.

> As far as I can tell from the azure docs it's not possible to use two different scheduling setting

Use the package build yml and also make debug builds of each package? That triggers only nightly, and sooner or later we may want to publish debug builds anyway (on a different conda label). Unless the debug build is published there is very limited use to it anyway, I suppose?

File: .azure-pipelines/package_nightly_and_release.yml [hunk 5]
```diff
@@ -47,6 +47,9 @@ extends:
       windows:
         py_versions: ['3.7', '3.8', '3.9']
         conda_env: 'scipp-developer-minimal.yml'
+      windows-debug:
+        py_versions: ['3.7', '3.8', '3.9']
+        conda_env: 'scipp-developer-minimal.yml'
```
Comment (SimonHeybrock) [inline .azure-pipelines/package_nightly_and_release.yml:52]: Which conda label will this deploy to?

Comment (Tom-Willemsen) [reply .azure-pipelines/package_nightly_and_release.yml:52]: As far as I can tell so far, all architectures are deploying into a conda label called `test`, unless I'm just missing where it's overridden. Does it make sense to create a `test-debug` label or similar to hold our debug builds?

Secondary question: do we also want nightly debug builds on linux/osx as well as windows?

Comment (SimonHeybrock) [reply .azure-pipelines/package_nightly_and_release.yml:52]: It should deploy to `dev` (see a couple lines below). But maybe we should split the build config discussion/changes from the fixes of the debug build? Can you remove all CI related changes from this PR, so we can merge the code fixes. We can then handle potential CI changes independently.

Comment (Tom-Willemsen) [reply .azure-pipelines/package_nightly_and_release.yml:52]: Ok, that's fine. I'll remove the CI changes now.

**Comment by SimonHeybrock** [comment]
@jl-wynen Have your comments been addressed?

**Comment by jl-wynen** [comment]
> @jl-wynen Have your comments been addressed?

Yes, well enough

**Review by jl-wynen — APPROVED** [review]
